src/main/scala/uk/gov/nationalarchives/consignmentexport/Validator.scala

```diff
@@ -0,0 +1,53 @@
+package uk.gov.nationalarchives.consignmentexport
+
+import java.time.LocalDateTime
+import java.util.UUID
+
+import cats.effect.IO
+import graphql.codegen.GetConsignmentExport.getConsignmentForExport.GetConsignment.Files
+import cats.implicits._
+import graphql.codegen.GetConsignmentExport.getConsignmentForExport.GetConsignment
+import graphql.codegen.GetConsignmentExport.getConsignmentForExport.GetConsignment.Files.Metadata
+import uk.gov.nationalarchives.consignmentexport.Validator.ValidatedFileMetadata
+
+class Validator(consignmentId: UUID) {
+  def validateConsignmentHasFiles(consignmentData: GetConsignment): IO[Files] = {
+    IO.fromOption(consignmentData.files.headOption)(new Exception(s"Consignment API returned no files for consignment $consignmentId"))
+  }
+
+  def validateConsignmentResult(consignmentResult: Option[GetConsignment]): IO[GetConsignment] = {
+    IO.fromOption(consignmentResult)(new RuntimeException(s"No consignment metadata found for consignment $consignmentId"))
+  }
+
+  def validateFileMetadataNotEmpty(filesList: List[Files]): IO[List[ValidatedFileMetadata]] = {
+    filesList.flatMap(file => {
+      file.metadata.productElementNames.zip(file.metadata.productIterator).collect(propertyNameToValue => propertyNameToValue._2 match {
+        case x: Option[_] if x.isEmpty => propertyNameToValue._1
+      }).toList match {
+        case head :: next => s"${file.fileId} is missing the following properties: ${(head :: next).mkString(", ")}".some
```

Comment:
I'm finding this error handling a bit hard to follow. I'd quite like to try to refactor it so it's easier to read, but I'd like to hear your thoughts on the IO vs Try/Either question above first, in case it affects how this works.

Reply:
I had a go at refactoring this code to help me understand it. Here's my suggested alternative:

```
  def validateFileMetadataNotEmpty(filesList: List[Files]): Either[RuntimeException, List[ValidatedFileMetadata]] = {
    val fileErrors: Seq[String] = filesList.flatMap(file => {
      val metadataPropertyNames = file.metadata.productElementNames
      val metadataValues = file.metadata.productIterator
      val missingPropertyNames = metadataPropertyNames.zip(metadataValues).filter(propertyNameToValue => {
        propertyNameToValue._2.isInstanceOf[None.type]
      }).map(_._1).toList

      missingPropertyNames match {
        case Nil => None
        case _ => s"${file.fileId} is missing the following properties: ${missingPropertyNames.mkString(", ")}".some
      }
    })
    
    fileErrors match {
      case Nil => Right(filesList.map(validatedMetadata))
      case _ => Left(new RuntimeException(fileErrors.mkString("\n")))
    }
  }

  private def validatedMetadata(f: Files): ValidatedFileMetadata = ValidatedFileMetadata(f.fileId,
    f.metadata.clientSideFileSize.get,
    f.metadata.clientSideLastModifiedDate.get,
    f.metadata.clientSideOriginalFilePath.get,
    f.metadata.foiExemptionCode.get,
    f.metadata.heldBy.get,
    f.metadata.language.get,
    f.metadata.legalStatus.get,
    f.metadata.rightsCopyright.get
  )
```

It does the same thing as before, but I've made a few changes which I think make it easier to follow:

* Split `validatedMetadata` into a separate method so you don't have to read it at the same time as the error handling code
* Add a few temporary variables to make the steps clearer, especially for things like `productElementNames` and `productIterator` - I don't know how common they are, but I'd never seen them before and had to look up what "product" meant in this context. I think doing lots of chaining works when each step is very clear, but when there's a lot of logic in each step then it's helpful to use variables to give each step a clear name.
* Used `case Nil` rather than `case head :: next` to identify empty lists, which I think is easier to follow
* Used `filter` rather than `collect`, which I think makes the intent clearer - I don't expect it will be a performance issue when there will only be up to thousands of files

What do you think about those?

I'd also suggest naming `validateFileMetadataNotEmpty` something like `extractFileMetadata`. The type signature makes it clear that it can return errors, and it does more than validation because it transforms the API representation of the metadata into a case class.

Reply:
That is better thanks, especially `case Nil`, I wasn't happy with `head :: next` but I couldn't find a way around it. I've renamed the method as well